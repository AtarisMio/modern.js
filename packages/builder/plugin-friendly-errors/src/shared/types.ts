import StackTracey from 'stacktracey';
import type * as webpack from 'webpack';

export type TracingFrame = StackTracey.Entry;

export interface ParsedError extends webpack.WebpackError {
  raw: webpack.WebpackError;
  trace: TracingFrame[];
}

export type ErrorTransformer = (error: ParsedError) => ParsedError | void;

export type ErrorFormatter = (error: ParsedError) => string | void;
