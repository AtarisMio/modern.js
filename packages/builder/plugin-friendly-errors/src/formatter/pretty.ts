import chalk from '@modern-js/utils/chalk';
import { ErrorFormatter } from '../shared/types';

export const prettyFormatter: ErrorFormatter = e => {
  const message = e.message || e;
  const name = e.name || 'Error';

  const formattedStack = e.trace
    .map(frame => {
      const { file, line, column } = frame;
      return `  at ${chalk.gray(file)}:${line}:${column}`;
    })
    .join('\n');

  return `${chalk.red(`${name}: ${message}`)}\n\n${formattedStack}`;
};
