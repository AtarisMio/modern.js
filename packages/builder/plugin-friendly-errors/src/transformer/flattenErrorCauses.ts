import _ from '@modern-js/utils/lodash';
import StackTracey from 'stacktracey';
import { ErrorTransformer } from '../shared/types';

export const compareStack = (
  stack1: readonly StackTracey.Entry[],
  stack2: readonly StackTracey.Entry[],
) => {
  if (stack1.length !== stack2.length) {
    return false;
  }
  for (let i = 0; i < stack1.length; i++) {
    const s1 = stack1[i];
    const s2 = stack2[i];
    if (s1.file !== s2.file || s1.line !== s2.line || s1.column !== s2.column) {
      return false;
    }
  }
  return true;
};

export const compareStackFragments = (
  stack1: readonly StackTracey.Entry[],
  stack2: readonly StackTracey.Entry[],
) => {
  const len = Math.min(stack1.length, stack2.length);
  return compareStack(
    stack1.length === len ? stack1 : stack1.slice(0, len),
    stack2.length === len ? stack2 : stack2.slice(0, len),
  );
};

export const mergeTraceHeads = (traces: StackTracey.Entry[][]) => {
  const heads: StackTracey.Entry[] = [];
  const tails: StackTracey.Entry[] = [];
  for (const [head, ...rest] of traces) {
    if (!heads.length) {
      heads.push(head);
      tails.push(...rest);
    } else if (compareStackFragments(tails, rest)) {
      heads.push(head);
    } else {
      return null;
    }
  }
  return [...heads, ...tails];
};

export const flattenErrorCauses: ErrorTransformer = e => {
  const traces = [e.trace, ..._.map(e.causes, 'trace')];
  const merged = mergeTraceHeads(traces);
  if (merged) {
    e.trace = merged;
    e.causes = [];
  }
  return e;
};
