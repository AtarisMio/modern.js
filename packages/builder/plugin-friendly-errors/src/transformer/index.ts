export * from './chunkRenderError';
export * from './moduleParseError';
