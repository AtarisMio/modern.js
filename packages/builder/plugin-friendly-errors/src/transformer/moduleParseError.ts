import chalk from '@modern-js/utils/chalk';
import StackTracey from 'stacktracey';
import { ErrorTransformer } from '../shared/types';

export const transformModuleParseError: ErrorTransformer = e => {
  if (e.name === 'ModuleParseError') {
    const rawError: Error = (e.raw as any).error;
    const rawStack = rawError?.stack;
    let sliceEnding = -1;
    rawStack && (sliceEnding += e.message.indexOf(rawStack));
    e.trace = new StackTracey(e.message).items;
    e.message = e.message.slice(0, sliceEnding);
    e.message += ' You can try to fix it by:\n';
    const tips = [
      'Check if the file is valid.',
      'Enable relational config of `tools`: https://modernjs.dev/builder/en/api/config-tools.html',
      'Install builder plugins: https://modernjs.dev/builder/en/plugin',
    ];
    e.message += tips.map(tip => ` ${chalk.gray(`*`)} ${tip}`).join('\n');
    e.message += '\n';
    e.message += 'Or you can try to configure bundler loaders manually.';
  }
};
