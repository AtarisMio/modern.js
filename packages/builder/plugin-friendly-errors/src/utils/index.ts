/**
 * Concat and flattens non-null values.
 * Ex: concat(1, undefined, 2, [3, 4]) = [1, 2, 3, 4]
 */
export function concat<T>(...args: T[]): T[] {
  const _args = Array.from(args).filter(e => e != null);
  const baseArray = Array.isArray(_args[0]) ? _args[0] : [_args[0]];
  return Array.prototype.concat.apply(baseArray, _args.slice(1));
}

/**
 * Dedupes array based on criterion returned from iteratee function.
 * Ex: uniqueBy(
 *     [{ id: 1 }, { id: 1 }, { id: 2 }],
 *     val => val.id
 * ) = [{ id: 1 }, { id: 2 }]
 */
export function uniqueBy<T>(arr: T[], func: (val: T) => any) {
  const seen: Record<any, number> = {};
  return arr.filter(el => {
    const e = func(el);
    return !(e in seen) && (seen[e] = 1);
  });
}
