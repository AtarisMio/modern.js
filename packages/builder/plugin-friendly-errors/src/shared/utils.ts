import assert from 'assert';
import { WebpackError } from 'webpack';
import StackTracey from 'stacktracey';
import { baseFormatter } from '../formatter/base';
import { ErrorFormatter, ErrorTransformer, ParsedError } from './types';

const _traceSerializeCache = new WeakMap<WebpackError, ParsedError>();

export const parseError = (error: WebpackError) => {
  if (_traceSerializeCache.has(error)) {
    return _traceSerializeCache.get(error)!;
  }

  const parsed = cloneObject(error) as ParsedError;
  parsed.trace = new StackTracey(error).items;
  parsed.raw = error;

  _traceSerializeCache.set(error, parsed);

  return parsed;
};

export const cloneObject = <T>(error: T): T => {
  const cloned = Object.create(Object.getPrototypeOf(error));
  Object.assign(cloned, error);
  return cloned;
};

export const transformError = (
  transformers: ErrorTransformer[],
  error: ParsedError,
): WebpackError => {
  return transformers.reduce(
    (error, transformer) => transformer(error) || error,
    error,
  );
};

export const formatError = (
  formatters: ErrorFormatter[] = [baseFormatter],
  error: ParsedError,
) => {
  let formatted: string | void;
  for (const formatter of formatters) {
    formatted = formatter(error);
    if (formatted) {
      break;
    }
  }
  assert(typeof formatted === 'string', 'No error formatter found');
  return formatted;
};
