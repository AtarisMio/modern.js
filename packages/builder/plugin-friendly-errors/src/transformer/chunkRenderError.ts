import { parseError } from '../shared/utils';
import { ErrorTransformer } from '../shared/types';

export const transformConnectAttachedHead: ErrorTransformer = e => {
  if (e.name === 'ChunkRenderError' && 'error' in e) {
    const attached = parseError((e as any).error);
    e.trace.splice(0, 0, attached.trace[0]);
  }
};
