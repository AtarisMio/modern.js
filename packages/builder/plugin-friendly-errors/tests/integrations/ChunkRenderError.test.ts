import assert from 'assert';
import { createStubBuilder } from '@modern-js/builder-webpack-provider/stub';
import { useFixture } from '@modern-js/e2e';
import { test } from 'vitest';
import { Chunk } from 'webpack';
import ChunkRenderError from 'webpack/lib/ChunkRenderError';
import { baseFormatter, prettyFormatter } from '@/formatter';
import { transformConnectAttachedHead } from '@/transformer';
import { parseError } from '@/shared/utils';

test('ChunkRenderError', async () => {
  const options = await useFixture('@modern-js/e2e/fixtures/builder/basic');
  const builder = await createStubBuilder({
    ...options,
    plugins: 'minimal',
  });
  const [{ stats }] = await builder.unwrapHook('onAfterBuildHook');
  assert(stats && 'compilation' in stats);
  const chunk = new Chunk('foo');
  const error = new ChunkRenderError(
    chunk,
    stats.compilation,
    new Error('foo'),
  );
  const parsed = parseError(error);
  console.log(baseFormatter(parsed));
  transformConnectAttachedHead(parsed);
  console.log(prettyFormatter(parsed));
});
