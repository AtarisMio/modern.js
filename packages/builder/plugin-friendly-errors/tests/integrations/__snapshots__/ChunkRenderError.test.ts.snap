// Vitest Snapshot v1

exports[`ChunkRenderError 1`] = `
"/chunkrendererror foo
    at /Users/bytedance/repositories/modern.js/packages/builder/plugin-friendly-errors/tests/integrations/ChunkRenderError.test.ts:28:17
    at async runTest (/Users/bytedance/repositories/modern.js/node_modules/<PNPM_INNER>/vitest/dist/chunk-runtime-error.87a2b5a2.mjs:488:5)
    at async runSuite (/Users/bytedance/repositories/modern.js/node_modules/.pnpm/vitest@0.21.1_@vitest+ui@0.21.1/node_modules/vitest/dist/chunk-runtime-error.87a2b5a2.mjs:576:13)
    at async runFiles (/Users/bytedance/repositories/modern.js/node_modules/.pnpm/vitest@0.21.1_@vitest+ui@0.21.1/node_modules/vitest/dist/chunk-runtime-error.87a2b5a2.mjs:620:5)
    at async startTestsNode (/Users/bytedance/repositories/modern.js/node_modules/.pnpm/vitest@0.21.1_@vitest+ui@0.21.1/node_modules/vitest/dist/chunk-runtime-error.87a2b5a2.mjs:638:3)
    at async /Users/bytedance/repositories/modern.js/node_modules/.pnpm/vitest@0.21.1_@vitest+ui@0.21.1/node_modules/vitest/dist/entry.mjs:76:9
    at async Module.withEnv (/Users/bytedance/repositories/modern.js/node_modules/.pnpm/vitest@0.21.1_@vitest+ui@0.21.1/node_modules/vitest/dist/chunk-runtime-error.87a2b5a2.mjs:259:5)
    at async run (/Users/bytedance/repositories/modern.js/node_modules/.pnpm/vitest@0.21.1_@vitest+ui@0.21.1/node_modules/vitest/dist/entry.mjs:71:5)
    at async file:/Users/bytedance/repositories/modern.js/node_modules/.pnpm/tinypool@0.2.4/node_modules/tinypool/dist/esm/worker.js:99:20
Error: foo
    at /Users/bytedance/repositories/modern.js/packages/builder/plugin-friendly-errors/tests/integrations/ChunkRenderError.test.ts:31:5
    at async runTest (/Users/bytedance/repositories/modern.js/node_modules/.pnpm/vitest@0.21.1_@vitest+ui@0.21.1/node_modules/vitest/dist/chunk-runtime-error.87a2b5a2.mjs:488:5)
    at async runSuite (/Users/bytedance/repositories/modern.js/node_modules/.pnpm/vitest@0.21.1_@vitest+ui@0.21.1/node_modules/vitest/dist/chunk-runtime-error.87a2b5a2.mjs:576:13)
    at async runFiles (/Users/bytedance/repositories/modern.js/node_modules/.pnpm/vitest@0.21.1_@vitest+ui@0.21.1/node_modules/vitest/dist/chunk-runtime-error.87a2b5a2.mjs:620:5)
    at async startTestsNode (/Users/bytedance/repositories/modern.js/node_modules/.pnpm/vitest@0.21.1_@vitest+ui@0.21.1/node_modules/vitest/dist/chunk-runtime-error.87a2b5a2.mjs:638:3)
    at async /Users/bytedance/repositories/modern.js/node_modules/.pnpm/vitest@0.21.1_@vitest+ui@0.21.1/node_modules/vitest/dist/entry.mjs:76:9
    at async Module.withEnv (/Users/bytedance/repositories/modern.js/node_modules/.pnpm/vitest@0.21.1_@vitest+ui@0.21.1/node_modules/vitest/dist/chunk-runtime-error.87a2b5a2.mjs:259:5)
    at async run (/Users/bytedance/repositories/modern.js/node_modules/.pnpm/vitest@0.21.1_@vitest+ui@0.21.1/node_modules/vitest/dist/entry.mjs:71:5)
    at async file:/Users/bytedance/repositories/modern.js/node_modules/.pnpm/tinypool@0.2.4/node_modules/tinypool/dist/esm/worker.js:99:20"
`;

exports[`ChunkRenderError 2`] = `
"[31mChunkRenderError: foo[39m

  at [90m/Users/bytedance/repositories/modern.js/packages/builder/plugin-friendly-errors/tests/integrations/ChunkRenderError.test.ts[39m:31:5
  at [90m/Users/bytedance/repositories/modern.js/packages/builder/plugin-friendly-errors/tests/integrations/ChunkRenderError.test.ts[39m:28:17
  at [90m/Users/bytedance/repositories/modern.js/node_modules/<PNPM_INNER>/vitest/dist/chunk-runtime-error.87a2b5a2.mjs[39m:488:5
  at [90m/Users/bytedance/repositories/modern.js/node_modules/.pnpm/vitest@0.21.1_@vitest+ui@0.21.1/node_modules/vitest/dist/chunk-runtime-error.87a2b5a2.mjs[39m:576:13
  at [90m/Users/bytedance/repositories/modern.js/node_modules/.pnpm/vitest@0.21.1_@vitest+ui@0.21.1/node_modules/vitest/dist/chunk-runtime-error.87a2b5a2.mjs[39m:620:5
  at [90m/Users/bytedance/repositories/modern.js/node_modules/.pnpm/vitest@0.21.1_@vitest+ui@0.21.1/node_modules/vitest/dist/chunk-runtime-error.87a2b5a2.mjs[39m:638:3
  at [90masync /Users/bytedance/repositories/modern.js/node_modules/.pnpm/vitest@0.21.1_@vitest+ui@0.21.1/node_modules/vitest/dist/entry.mjs[39m:76:9
  at [90m/Users/bytedance/repositories/modern.js/node_modules/.pnpm/vitest@0.21.1_@vitest+ui@0.21.1/node_modules/vitest/dist/chunk-runtime-error.87a2b5a2.mjs[39m:259:5
  at [90m/Users/bytedance/repositories/modern.js/node_modules/.pnpm/vitest@0.21.1_@vitest+ui@0.21.1/node_modules/vitest/dist/entry.mjs[39m:71:5
  at [90masync file:/Users/bytedance/repositories/modern.js/node_modules/.pnpm/tinypool@0.2.4/node_modules/tinypool/dist/esm/worker.js[39m:99:20"
`;
